main.py

```diff
@@ -1,238 +1,275 @@
+    def __init__(self):
+        super().__init__()
+        self._tokens = []
+
+    def on_llm_new_token(self, token: str, **kwargs: Any) -> None:
+        self._tokens.append(token)
+
+    def on_llm_end(self, response: LLMResult, *, run_id: UUID, parent_run_id: UUID | None = None, **kwargs: Any) -> Any:
+        # Calculate the cost based on OpenAI's pricing
+        cost_per_token = ((len(self._tokens) / 1000) * 0.002) * 83.33  # INR Cost per token
+        total_tokens = len(self._tokens)
+        print(f"\n\nTotal tokens generated: {total_tokens}")
+        print(f"Total cost: {cost_per_token} INR")
+
+# Function to load the LLM 
+def load_model(device_type:str = DEVICE_TYPE, model_id:str = MODEL_NAME, model_basename:str = MODEL_FILE, LOGGING=logging, quantized_model_id:str = None):
+    """
+        input: device_type, model_id, model_basename, LOGGING
+        output: Hugging Face model
+        description:  The function loads the pre-trained model from Hugging Face and returns the loaded model.
+    """
+    if model_basename is not None and ".gguf" in model_basename.lower() :
+
+        callback_manager = CallbackManager([StreamingStdOutCallbackHandler(),TokenCallbackHandler()])
+        try:
+        # Download the model checkpoint from the Hugging Face Hub
+            model_path = hf_hub_download(
+                repo_id=model_id,
+                filename=model_basename,
+                resume_download=True,
+                cache_dir=MODEL_DIRECTORY,
+            )
+            # Model Parameters
+            kwargs = {
+                "model_path": model_path,
+                "max_tokens": MAX_TOKEN_LENGTH,
+                "n_ctx": MAX_TOKEN_LENGTH,
+                "n_batch": 512,  
+                "callback_manager": callback_manager,
+                "verbose":False,
+                "f16_kv":True,
+                "streaming":True,
+            }
+            if device_type.lower() == "mps":
+                kwargs["n_gpu_layers"] = 1 # only for MPS devices
+            if device_type.lower() == "cuda":
+                kwargs["n_gpu_layers"] = N_GPU_LAYERS  # set this based on your GPU
+            # Create a LlamaCpp object (language model)
+            llm =  LlamaCpp(**kwargs)
+            LOGGING.info(f"Loaded {model_id} locally")
+            return llm  # Returns a LlamaCpp object (language model)
+        except Exception as e:
+            LOGGING.info(f"Error {e}")
+
+    elif model_basename is not None and ".bin" in model_basename.lower() :
+        try:
+            quantize_config = BaseQuantizeConfig(
+                bits=4,  # quantize model to 4-bit
+                group_size=128,  # it is recommended to set the value to 128
+                desc_act=True,  # set to False can significantly speed up inference but the perplexity may slightly bad
+            )
+            # Load the Hugging Face model into CPU memory (default)
+            if (device_type == "cpu"):

```

Comment:
You can make it to be dynamic since DEVICE_TYPE is a global variable that can be used and passed as parameter

Reply:
You can just use model_id itself rather than adding another variable to the function

Reply:
i think quantized models should specified using a different variable ?

Reply:
Based on the file type we can classify them

Reply:
In future we can put the different types of models as seperate functions. And adding modularity to the code. Since the main.py is growing big as we add more functionality to the project

Reply:
@neokd Sorry, what do you mean by dynamic ? the default is to use CPU for inference. Or as an alternative first device_type could be checked for GPU else go with the CPU inference method.

Reply:
and one more thing, is `.bin` correct for GPTQ models ?

Reply:
Dynamic mean there is already an global variable called DEVICE_TYPE  which detects the device type and sets it accordingly.

Reply:
So should `AutoGPTQForCausalLM.from_quantized` be dropped ? and instead use CPU inference always ?

Reply:
> config

This would be okay. Actually I got confused with the device_type. Sorry 😅😅

Reply:
If there are any more things to change, do let me know

Reply:
No everything is fine.I'll review it and merge it completely